calico_node/tests/st/config/test_felix_config.py

```diff
@@ -0,0 +1,410 @@
+    """
+    Tests felix configurations setup by calicoctl.
+    """
+    hosts = None
+
+    @classmethod
+    def setUpClass(cls):
+        # Wipe etcd once before any test in this class runs.
+        _log.debug("Wiping etcd")
+        wipe_etcd(HOST_IPV4)
+
+        # Test felix configurations.
+
+        # Create two hosts.
+        cls.hosts = []
+        cls.host1 = DockerHost("cali-host1",
+                               additional_docker_options=CLUSTER_STORE_DOCKER_OPTIONS,
+                               post_docker_commands=POST_DOCKER_COMMANDS,
+                               start_calico=False)
+        cls.host1_hostname = cls.host1.execute("hostname")
+        cls.host2 = DockerHost("cali-host2",
+                               additional_docker_options=CLUSTER_STORE_DOCKER_OPTIONS,
+                               post_docker_commands=POST_DOCKER_COMMANDS,
+                               start_calico=False)
+        cls.host2_hostname = cls.host2.execute("hostname")
+        cls.hosts.append(cls.host1)
+        cls.hosts.append(cls.host2)
+
+        # Start calico node on hosts.
+        for host in cls.hosts:
+            host.start_calico_node()
+
+        _log.info("host1 IP: %s , host2 IP: %s", cls.host1.ip, cls.host2.ip)
+
+        clear_on_failures()
+        add_on_failure(cls.host1.log_extra_diags)
+        add_on_failure(cls.host2.log_extra_diags)
+
+    @handle_failure
+    def test_failsafe_inbound(self):
+        """
+        Test failsafe inbound configuration.
+        """
+        random = {"udp": [70], "tcp": [3000, 5000]}
+        default = {"udp": [68], "tcp": [22, 179, 2379, 2380, 6666, 6667]}
+
+        # Test random and default ports can be accessed with no host endpoints setup.
+        self.run_failsafe(default, True)
+        self.run_failsafe(random, True)
+
+        # With host endpoints setup. Default ports can be accessed.
+        # But random ports can not.
+        self.add_host_iface(self.host1_hostname, self.host1.ip)
+        self.run_failsafe(default, True)
+        self.run_failsafe(random, False)
+
+        # Update felix config and check again.
+        self.add_felix_failsafe_config("default", random, "inbound")
+        self.run_failsafe(default, False)
+        self.run_failsafe(random, True)
+
+        # Put default back for host1.
+        self.add_felix_failsafe_config("node.%s" % self.host1_hostname, default, "inbound")
+        self.run_failsafe(default, True)
+        self.run_failsafe(random, False)
+
+        # Put default back for all.
+        self.add_felix_failsafe_config("default", default, "inbound")
+        return
+
+    @handle_failure
+    def test_failsafe_outbound(self):
+        """
+        Test failsafe outbound configuration.
+        """
+        # We need to put 2379 into random because felix need to connect to etcd.
+        random = {"udp": [70], "tcp": [2379, 3000, 5000]}
+        default = {"udp": [53, 67], "tcp": [179, 2379, 2380, 6666, 6667]}
```

Comment:
How does it work that 2379 is in both TCP lists?

Reply:
If 2379 is not in random failsafe port list, as soon as we switch over to random failsafe ports, felix is not able to connect to etcd and could not continue test because it cannot receive any datastore updates.

Reply:
Does `run_failsafe` know not to check 2379 when its expecting the ports to fail?

Reply:
OMG. It is a miracle. :) How come they (2379: True, 2379: False) both pass one after another.

Reply:
Miracle or bug, one or the other :)